WIP make dynamic tabs follow ARIA 1.1 practices

patrickhlauke opened the pull request:
Closes https://github.com/twbs/bootstrap/issues/28918

patrickhlauke committed:
Add initial tabindex handling for tabs

set `tabindex` as well as `aria-selected` on tabs

patrickhlauke committed:
Add initial tabindex handling for tabs

set `tabindex` as well as `aria-selected` on tabs

patrickhlauke committed:
WIP add initial keyboard handling and initialisation

- initialisation happens very inelegantly - could do with some advice on how to move this to a function (should it be public, private, static?)
- keyboard handling works, but seems to get confused when switching too quickly between tabs sometimes
- activating tabs currently done via click() - there must be a nicer way to do this...how? fire _activate? which parameters to pass?

patrickhlauke committed:
WIP add initial keyboard handling and initialisation

- initialisation happens very inelegantly - could do with some advice on how to move this to a function (should it be public, private, static?)
- keyboard handling works, but seems to get confused when switching too quickly between tabs sometimes
- activating tabs currently done via click() - there must be a nicer way to do this...how? fire _activate? which parameters to pass?

patrickhlauke commented:
WIP only at this stage, as i'd need some help with two particular aspects:

- how to properly run the initialisation code on page load (which sanitizes any tablists on the current page, to ensure they have correct tabindex values, and that one item is always the active/selected one)
- how to actually call the tab activation from inside the keyboard handler - currently doing naive `click()` but there must be a better way (particularly since it seems if moving too quickly between tabs the script doesn't fire correctly and you may end up with two tab panels showing at the same time)

@Johann-S or others...any pointers? (and yes, no unit tests yet...they'll come)

patrickhlauke committed:
Remove dropdown-in-tabs related code

- there's no ARIA pattern that would allow dropdowns in tablists; it cannot be expressed accessibly to AT users. it also has serious usability drawbacks. we're already saying in the docs it should not be
used ... this goes a step further and removes the handling for it

patrickhlauke committed:
Remove dropdown-in-tabs related code

- there's no ARIA pattern that would allow dropdowns in tablists; it cannot be expressed accessibly to AT users. it also has serious usability drawbacks. we're already saying in the docs it should not be
used ... this goes a step further and removes the handling for it

patrickhlauke committed:
Add explicit note about aria-orientation, add all missing role="presentation"

`role="presentation"` is needed as otherwise AT (like NVDA) will get confused/won't be able to "count" the total number of tabs properly

patrickhlauke committed:
Add explicit note about aria-orientation, add all missing role="presentation"

`role="presentation"` is needed as otherwise AT (like NVDA) will get confused/won't be able to "count" the total number of tabs properly

patrickhlauke committed:
Merge branch 'master' into patrickhlauke-tabs-kbd-automatic-activation

patrickhlauke commented:
https://deploy-preview-28927--twbs-bootstrap.netlify.com/docs/4.3/components/navs/#javascript-behavior

- only one (the selected/active) tab control can be focused using keyboard
- cursor keys left/right change tabs
- tabs auto-activate (following the "Tabs with automatic activation" https://www.w3.org/TR/wai-aria-practices-1.1/examples/tabs/tabs-1/tabs.html and https://www.w3.org/TR/wai-aria-practices-1.1/#tabpanel / pattern)
- `aria-orientation="vertical"` uses cursor up/down
- removed the code for dropdowns inside tabs, as it's not a sensible/supported pattern for ARIA and has issues, so might as well remove it...this is arguably a breaking change (though from previous question to devs, nobody seemed to really use it https://github.com/twbs/bootstrap/issues/23035)

Johann-S commented on js/src/tab.js:
```diff
@@ -229,6 +217,48 @@ class Tab {
 
+  static _dataApiKeydownHandler(event) {
+    const tablist = SelectorEngine.closest(event.target, Selector.TABLIST)
+    let tablistorientation = tablist.getAttribute('aria-orientation')
```
use `const` here and maybe a bit of camel case, for example: `const tabListOrientation`

patrickhlauke replied:
but I'm potentially changing the value of `tablistorientation` in the next bit, so not a constant?

Johann-S replied:
in that case `let` if fine 👍

Johann-S commented on js/src/tab.js:
```diff
@@ -229,6 +217,48 @@ class Tab {
+  static _dataApiKeydownHandler(event) {
+    const tablist = SelectorEngine.closest(event.target, Selector.TABLIST)
+    let tablistorientation = tablist.getAttribute('aria-orientation')
+    if (tablistorientation !== 'vertical') {
```
We should store tabs orientation in something like:
```
const Orientation = {
	VERTICAL: 'vertical',
	HORIZONTAL: 'horizontal'
}

Johann-S commented on js/src/tab.js:
```diff
@@ -240,6 +270,36 @@ class Tab {
 
+EventHandler.on(window, Event.LOAD_DATA_API, () => {
+  const tablists = makeArray(SelectorEngine.find(Selector.TABLIST))
+  if (tablists.length === 0) {
```
`if (!tablists.length)`

Johann-S commented on js/src/tab.js:
```diff
@@ -240,6 +270,36 @@ class Tab {
+    let selectedTabFound = false
+
+    // iterate over each tab in the tablist, make sure they have correct tabindex/aria-selected
+    for (let j = 0; j < tabs.length; j++) {
```
just a suggestion but you can use `.forEach` too

patrickhlauke replied:
still old-school me, so not quite sure how to remunge the code...particularly since at the end I jump back to the very first tab in the array if no selected tab was found during the run.

Johann-S replied:
something like: 
```js
tabs.forEach((tab) => {
  if (tab.getAttribute('aria-selected') === 'true' && !selectedTabFound) {
    tab.setAttribute('tabindex', '0')
    selectedTabFound = true
  } else {
    tab.setAttribute('tabindex', '-1')
    tab.setAttribute('aria-selected', 'false')
  }
})
```

btw I added some changes, it's not needed to use `.hasAttribute` because if no attribute `.getAttribute` will return `null`.

Johann-S commented on js/src/tab.js:
```diff
@@ -240,6 +270,36 @@ class Tab {
+    }
+
+    // if none of the tabs were explicitly marked as selected, pick first one
+    if (selectedTabFound === false) {
```
`if (!selectedTabFound)`

Johann-S commented:
Hi @patrickhlauke,

About your use of `.click()` for me the main issue is when the page is loaded you do not create any Tab instance, you just add accessibility attributes, you should initialize Tabs with: `const tab = new Tab(element)` and after that on your keyboard listener you'll be able to do: `Data.getData(element)` to get a Tab instance.
Like that you'll be able to call any public methods of our Tab component.

I hope what I said is clear 😟 if not, do not hesitate to ask me 😃

patrickhlauke commented:
> About your use of `.click()` for me the main issue is when the page is loaded you do not create any Tab instance, you just add accessibility attributes, you should initialize Tabs with: `const tab = new Tab(element)` and after that on your keyboard listener you'll be able to do: `Data.getData(element, DATA_KEY)` to get a Tab instance.
> Like that you'll be able to call any public methods of our Tab component.
> 
> I hope what I said is clear 😟 if not, do not hesitate to ask me 😃

I'll admit that I have no clue what all our magic instances and scripts and DATA_KEY things etc actually do at this point...I just cobble together bits from other things that seem to work. I'll have another crack at this tonight though to see if I can work it out...

As for the initialisation...does it make sense to say it should be in a separate function, and to call that rather than cramming it all into the event listener definition there? he main thing for my page initialisation is: it needs to run on page load before the user even starts interacting with any tab interfaces, to make sure the correct tabindices and `aria-selected` are set (so before any keyboard listener is even called). But not quite sure how to achieve that with all the code BS uses (I was trying some basic approaches, but kept getting various weird errors with my naive approaches). Maybe if you have a basic gist of how it could/should look that could give me a clue...

(that's partly the problem for me with these things...give me basic vanilla JS, and I can probably spaghetti code my way to the solution...but with the various abstractions like SelectorEngine etc that we use in BS, I'm mostly lost now...)

patrickhlauke committed:
Initial code tweaks based on comments

patrickhlauke committed:
Initial code tweaks based on comments

patrickhlauke committed:
Remove dropdown-in-tabs related unit tests

Since that logic has now been expunged

patrickhlauke committed:
Remove dropdown-in-tabs related unit tests

Since that logic has now been expunged

patrickhlauke committed:
Make note about dropdowns in tabs an actual callout

patrickhlauke committed:
Make note about dropdowns in tabs an actual callout

patrickhlauke committed:
Replace for loop with forEach

patrickhlauke committed:
Replace for loop with forEach

XhmikosR commented on js/src/tab.js:
```diff
@@ -287,15 +287,15 @@ EventHandler.on(window, Event.LOAD_DATA_API, () => {
-    for (let j = 0; j < tabs.length; j++) {
-      if (tabs[j].getAttribute('aria-selected') === 'true' && !selectedTabFound) {
-        tabs[j].setAttribute('tabindex', '0')
+    tabs.forEach(tab => {
```
You should use our helper function `makeArray` here.

patrickhlauke replied:
any pointers of how to do that would be welcome. as said, i've got very little idea of the idiosyncracies of our helper functions, admittedly...

XhmikosR replied:
You can look at how it's already being used like here https://github.com/twbs/bootstrap/blob/master/js/src/carousel.js#L337

patrickhlauke replied:
tried to tighten that into `makeArray(SelectorEngine.find(Selector.DATA_TOGGLE, tablists[i])).forEach(tab => {` but then not obvious how after the loop i can revisit the first tab to set it unless i run some other SelectorEngine thing again, which seems wasteful/pointless?

XhmikosR replied:
There's a reason we have this utility. Some browsers don't support `NodeList.forEach`

patrickhlauke replied:
but i am using makeArray a few lines earlier...not quite sure i follow what you're saying here?

patrickhlauke replied:
as in `tabs` is defined a few lines earlier already as `const tabs = makeArray(SelectorEngine.find(Selector.DATA_TOGGLE, tablists[i]))` ... just that i'm not currently chaining the `forEach` directly after it but storing it in the variable

XhmikosR replied:
Ah, nvm then, it should be fine.

XhmikosR commented:
You can look at how it's already being used like here https://github.com/twbs/bootstrap/blob/master/js/src/carousel.js#L337

patrickhlauke committed:
Tweak

Stick with the naive `tabs[index].click()` as it works correctly, and trying to trigger just `show()` of a tab instance resulted in weird side effects...

Tighten the outer `makeArrary`/`forEach` loop (assuming when there's an empty array the forEach just bails, so no need to check the length?)

patrickhlauke committed:
Tweak

Stick with the naive `tabs[index].click()` as it works correctly, and trying to trigger just `show()` of a tab instance resulted in weird side effects...

Tighten the outer `makeArrary`/`forEach` loop (assuming when there's an empty array the forEach just bails, so no need to check the length?)

patrickhlauke commented:
Think at this stage i got to the end of my understanding/knowledge of how to tackle this for now. @XhmikosR or @Johann-S  ... not sure if you have time or inclination to point me to the most sensible way to do the initialization (rather than cramming it all directly into the load event listener thing)? Or is that actually ok? (looks ugly to me)

additionally, one aspect that is still broken: for tab panels with a `fade`, it seems that using keyboard i can switch tabs quicker than the fade actually happening/completing, leading to situations where a new tab is being shown while the one that was in the process of being shown isn't fully shown yet...and you end up with two tab panels showing at the same time. again, any pointer/indication of how this could be tackled would be appreciated...

and yes, also need to come up with unit tests to check the new keyboard behavior, and to check the initialisation happens properly...

patrickhlauke committed:
Merge branch 'master' into patrickhlauke-tabs-kbd-automatic-activation

Johann-S commented:
I don't think it's ugly to initialize our tabs in the load event, we do that for our other components for example Carousel or Scrollspy, so feel free to do it that way.

I don't have an answer for your other issue, but I'll take a look into it when I'll find the time

patrickhlauke commented:
if you pardon the enormous GIF for a second, here's an illustration of the problem when switching tabs too quickly with keyboard...

![skillsoft-learning-transcript-animated-focus-demonstration](https://user-images.githubusercontent.com/895831/59766027-38aba200-9297-11e9-97e3-d2973e82f109.gif)

patrickhlauke committed:
Add missing aria-orientation to list group tablists

For correct keyboard handling (cursor up/down) when interacting with these

patrickhlauke committed:
Add missing aria-orientation to list group tablists

For correct keyboard handling (cursor up/down) when interacting with these

patrickhlauke committed:
Merge branch 'master' into patrickhlauke-tabs-kbd-automatic-activation

patrickhlauke committed:
Merge branch 'master' into patrickhlauke-tabs-kbd-automatic-activation

patrickhlauke committed:
Merge branch 'master' into patrickhlauke-tabs-kbd-automatic-activation

patrickhlauke committed:
Merge branch 'master' into patrickhlauke-tabs-kbd-automatic-activation

patrickhlauke commented:
I've tried tackling this problem, but my understanding of how the whole async stuff is happening is preventing me from making any progress. In essence, the problem is due to the `.fade` and the assumption the code seems to make that there's only ever one previous tab that is currently still transitioning. when using a mouse, it seems you're never faster than the transition in clicking (or it somehow just ignores the click?), but when using the keyboard you can fire the tab changes very quickly. it further happens specifically when you do a quick "left, then right" with the cursor keys, i.e. you leave a tab, the previous active shown panel begins to fade out, the new panel begins to fade in, but then you're going back to the previous tab. somewhere along the line the code gets confused and then BOTH those tabs are shown. @Johann-S @XhmikosR any help on this would be appreciated, as apart from this stupid fading behavior, everything else seems to work just great (barring the absence of unit tests, which i'll tackle once we get this all working properly)...

Johann-S committed:
add transitioning state to avoid displaying two tabs in the same time

Johann-S commented:
just a question @redux, is it possible for a `.tab-content` to have `role="tablist"` ? Or is it just for the list of tabs

patrickhlauke commented:
> just a question @Redux, is it possible for a `.tab-content` to have `role="tablist"` ? Or is it just for the list of tabs

the latter. it's only for the list of the tabs/triggers. the content panels need to be outside of the tablist. (that was an old bug in ARIA 1.0 that was amended in 1.1)

patrickhlauke committed:
Merge branch 'master' into patrickhlauke-tabs-kbd-automatic-activation

patrickhlauke commented:
just checked the deployed preview of your change @Johann-S ... YAY, seems to work fine. i'll polish this a bit more and add unit tests etc, but this fills me with joy :)

Johann-S commented:
The main reason was because when I worked on your branch, I found an error in `tabs.html` in visual tests, and it was because one tab content had `role="tablist"` 😄 

You're welcome 😉

Johann-S committed:
separate file for our polyfills to have lighter plugins

patrickhlauke committed:
Merge branch 'master' into patrickhlauke-tabs-kbd-automatic-activation

mdo committed:
Fixes #28994

XhmikosR committed:
Fix typo. (#29008)

MartijnCuppens committed:
Remove attribute selectors (#28988)

ysds committed:
Drop support for .form-control-plaintext inside .input-group (#28972)

* Revert "Fix readonly-plain-text with button addon (#25871)"

This reverts commit ff40e00323f9256d8dfb678d13b57bfb01ac7354.

* Add migration comment

mdo committed:
v5: Forms update (#28450)

* Initial spike of consolidated form checks

* Stub out forms rearrangement

- Prepping to drop non-custom file and range inputs
- Prepping to merge custom and native checks and radios (with switches)
- Prepping to merge custom select with form select
- Moving docs arround so forms has it's own area given volume of CSS

* Move input group Sass file to forms subdir

* Start to split and move the docs around

* Simpler imports

* Copyediting

* delete overview file

* Remove commented out code

* remove the custom-forms import

* rewrite flex-check as form-check, replace all custom properties

* Remove old forms doc

* stub out new subpage link section

* update migration guide

* Update nav, forms overview in page nav, and descriptions

* fix check bg position

* fix margin-top calculation

* rename .custom-select to .form-select

* Update validation styles for new checks

* add some vertical margin, fix inline checks

* fix docs examples

* better way to do this contents stuff, redo the toc while i'm at it

* page restyle for docs while here

* un-callout that, edit text

* redo padding on toc

* fix toc

* start to cleanup checks docs

* Rewrite Markdown tables into HTML

* Redesign tables, redo their docs

* Replace Open Iconic icons with custom Bootstrap icons

* Redesign the docs navbar, add a subheader, redo the sidebar

* Redesign docs homepage a bit

* Simplify table style overrides for docs tables

* Simplify docs typography for page titles and reading line length

* Stub out icons page

* Part of sidebar update, remove migration from nav.yml

* Move toc CSS to separate partial

* Change appearance of overview page

* fix sidebar arrow direction

* Add footer to docs layout

* Update descriptions

* Drop the .form-group class for margin utilities

* Remove lingering form-group-margin-bottom var

* improve footer spacing

* add headings to range page

* uncomment form range css

* Rename .custom-range to .form-range

* Drop unused docs var

* Uncomment the comment

* Remove unused variable

* Fix radio image sizing

* Reboot update: reset horizontal ul and ol padding

* de-dupe IDs

* tweak toc styles

* nvm, fix dropdown versions stuff

* remove sidebar nav toggle for now

* broken html

* fix more broken html, move css

* scss linting

* comment out broken helper docs

* scope styles

* scope styles

* Fixes #25540 and fixes #26407 for v5 only

* Update sidebar once more

* Match new sidenav order

* fix syntax error

* Rename custom-file to form-file, update paths, update migration docs for previous changes in #28696

* rename back

* fix size and alignment

* rename that back too

mdo committed:
dist v5

Brianmanden committed:
Update collapse.md (#29025)

Corrected minor spelling error.

patrickhlauke committed:
Merge branch 'patrickhlauke-tabs-kbd-automatic-activation' of https://github.com/twbs/bootstrap into patrickhlauke-tabs-kbd-automatic-activation

patrickhlauke committed:
Unit test for tablist initialization

- check that `tabindex` and `aria-selected` are properly set up with only one `aria-selected="true"` and `tabindex="0"`
- check that as fallback the first tab control is set to `aria-selected="true"`

patrickhlauke committed:
Fix one of the existing unit tests

the markup fragment does not reflect proper code structure (surprising that it still worked...)

patrickhlauke committed:
Remove redundant sanity checks

unnecessary, as these checks should never fail (since the handler would not be called if there weren't any tabs, and the `index > 0` clause in the preceding `if` ensures that `index` will never be negative

patrickhlauke committed:
Add extensive keyboard handling unit tests

very verbose, but fully tests that cursor keys are handled correctly and have the desired final result (though stops short of checking the actual tab panels themselves)

patrickhlauke committed:
Fix visual unit test

fix the borked examples of dynamic tab panels, and remove the now unsupported tabs with dropdowns

patrickhlauke commented:
last remaining (non-trivial) work here is to add some sensible handling of disabled tabs. they should be skipped/ignored (e.g. if i have 3 tabs, and the second one is disabled, then when focused on the first one and pressing right cursor key, the third tab should be focused/become active)

patrickhlauke committed:
Merge branch 'master' into patrickhlauke-tabs-kbd-automatic-activation

Johann-S commented:
Hey @patrickhlauke, I'm sure you'll hate me 😆 but I removed QUnit and switch to Jasmine, so now your unit tests should go there: https://github.com/twbs/bootstrap/blob/master/js/src/tab/tab.spec.js

`js/tests/unit/tab.js` have been removed, so if you need any help do not hesitate 😉

patrickhlauke commented:
heh, the dangers of taking too long / not having time to get this over the line. no worries, i'll have to look at this jasmine thing eventually anyway, might as well do it as part of this ;)

patrickhlauke commented:
would love to review efforts along these lines...at this point, probably worth scrapping this PR and starting from scratch again. but my lack of JS chops will probably still raise its head once I hit the trickier parts. so preemptively asking: anybody (@Johann-S, others?) able to collaborate on a new take? I can explain what things should do, but once it gets to integrating that code and how all the event model in BS works...

patrickhlauke commented:
putting out a call again for anybody who'd have time/inclination to help take this forward with me and help me overcome my woefully lacking JS skills...

XhmikosR commented:
@patrickhlauke if you could make this from scratch against main, it should make things easier to look into.

XhmikosR committed:
Update tab.js

patrickhlauke commented:
Superseded by https://github.com/twbs/bootstrap/pull/33079
